File: tools/dsl-harness/kernels/triton/zig/fused_moe.zig
//! Registration of `fused_moe_kernel` for the harness.
//!
//! Re-exports the production Kernel from `zml.moe.triton_kernels`
//! (`FusedMoe`). One default sweep matching the bf16 / no-quant / no-bias
//! path that `zml.moe.fusedExpertsImpl` actually launches (mirrors the
//! `withConfig(...)` in `examples/triton_emitter/kernels_zig.zig`).
//!
//! Known divergence vs. the Python reference (`moe.py:fused_moe_kernel`,
//! vLLM-derived): the Python source loads every `stride_*_ptr` (incl. the
//! quant ones) unconditionally, while this DSL port only loads the strides
//! the bf16 path uses. The harness diff will surface that.

const std = @import("std");

const harness = @import("harness");
const zml = @import("zml");
const ops = zml.ops;
const Tensor = zml.Tensor;
const Shape = zml.Shape;

pub const Kernel = zml.moe.triton_kernels.FusedMoe.Kernel;

pub const SWEEPS: []const harness.Sweep(Kernel.Config) = &.{
    .{ .label = "default", .cfg = .{
        .a_dtype = .bf16,
        .b_dtype = .bf16,
        .c_dtype = .bf16,
        .a_scale_dtype = null,
        .b_scale_dtype = null,
        .b_bias_dtype = null,
        .topk_weights_dtype = null,
        .block_size_m = 64,
        .block_size_n = 64,
        .block_size_k = 32,
        .group_size_m = 4,
        .top_k = 2,
        .naive_block_assignment = false,
        .mul_routed_weight = true,
        .compute_type = .bf16,
        .use_fp8_w8a8 = false,
        .use_int8_w8a8 = false,
        .use_int8_w8a16 = false,
        .per_channel_quant = false,
        .has_bias = false,
    } },
};

// =============================================================================
// XLA-pipeline driver. 24 inputs + 1 output (`c`) → 25 `tt.func` params.
// Shapes are synthetic placeholders; XLA only runs the codegen pipeline.
// =============================================================================

const M: i64 = 256;
const N: i64 = 1024;
const K: i64 = 1024;
const E: i64 = 8;

threadlocal var active_ttir: [:0]const u8 = "";

pub fn setActiveTtir(ttir: [:0]const u8) void {
    active_ttir = ttir;
}

pub fn forward(
    a: Tensor, b: Tensor, b_bias: Tensor, a_scale: Tensor, b_scale: Tensor,
    topk_weights: Tensor, sorted_token_ids: Tensor, expert_ids: Tensor, num_tokens_post_padded: Tensor,
    n_ptr: Tensor, k_ptr: Tensor, em_ptr: Tensor, num_valid_tokens: Tensor,
    stride_am: Tensor, stride_be: Tensor, stride_bn: Tensor, stride_cm: Tensor,
    stride_asm: Tensor, stride_ask: Tensor, stride_bse: Tensor, stride_bsk: Tensor, stride_bsn: Tensor,
    stride_bbe: Tensor, stride_bbn: Tensor,
    _: Tensor,
) Tensor {
    return ops.triton(
        .{
            a, b, b_bias, a_scale, b_scale,
            topk_weights, sorted_token_ids, expert_ids, num_tokens_post_padded,
            n_ptr, k_ptr, em_ptr, num_valid_tokens,
            stride_am, stride_be, stride_bn, stride_cm,
            stride_asm, stride_ask, stride_bse, stride_bsk, stride_bsn,
            stride_bbe, stride_bbn,
        },
        .{Shape.init(.{ M, N }, .bf16)},
        .{
            .name = Kernel.name,
            .ir = active_ttir,
            .grid = .{ 16, 1, 1 },
            .num_warps = 4,
            .num_stages = 2,
        },
    )[0];
}

pub fn args() std.meta.ArgsTuple(@TypeOf(forward)) {
    const p = struct {
        fn t() Tensor {
            return Tensor.init(.{1}, .i64);
        }
    }.t;
    return .{
        Tensor.init(.{ M, K }, .bf16), // a_ptr
        Tensor.init(.{ E, N, K }, .bf16), // b_ptr
        Tensor.init(.{ E, N }, .bf16), // b_bias_ptr
        Tensor.init(.{1}, .f32), // a_scale_ptr
        Tensor.init(.{1}, .f32), // b_scale_ptr
        Tensor.init(.{M}, .f32), // topk_weights_ptr
        Tensor.init(.{M}, .i32), // sorted_token_ids_ptr
        Tensor.init(.{ E * 4 }, .i32), // expert_ids_ptr
        Tensor.init(.{1}, .i32), // num_tokens_post_padded_ptr
        p(), p(), p(), p(), // N_ptr, K_ptr, EM_ptr, num_valid_tokens_ptr
        p(), p(), p(), p(), // stride_am_ptr, stride_be_ptr, stride_bn_ptr, stride_cm_ptr
        p(), p(), p(), p(), p(), // stride_asm_ptr, stride_ask_ptr, stride_bse_ptr, stride_bsk_ptr, stride_bsn_ptr
        p(), p(), // stride_bbe_ptr, stride_bbn_ptr
        Tensor.init(.{ M, N }, .bf16), // c_ptr placeholder
    };
}

File: tools/dsl-harness/kernels/triton/py/fused_moe.py
"""Python reference for `fused_moe_kernel` (bf16 / no-quant / no-bias path)."""

from __future__ import annotations

from typing import Any, Dict, Tuple

import triton.language as tl

from triton_helpers import dtype_str, fake as _t
from moe import fused_moe_kernel, write_zeros_to_output  # type: ignore  # noqa: F401

__all__ = ["fused_moe_kernel", "build_args"]

_TL_DTYPES = {"bf16": tl.bfloat16, "fp16": tl.float16, "fp32": tl.float32}

# Synthetic shape budget — XLA never launches; sizes are placeholders.
_M = 256
_N = 1024
_K = 1024
_E = 8


def build_args(cfg: Dict[str, Any]) -> Tuple[list, Dict[str, Any]]:
    a_dtype = dtype_str(cfg.get("a_dtype"), "bf16")
    b_dtype = dtype_str(cfg.get("b_dtype"), "bf16")
    c_dtype = dtype_str(cfg.get("c_dtype"), "bf16")
    a_scale_dtype = dtype_str(cfg.get("a_scale_dtype"), "fp32")
    b_scale_dtype = dtype_str(cfg.get("b_scale_dtype"), "fp32")
    b_bias_dtype = dtype_str(cfg.get("b_bias_dtype"), c_dtype)
    topk_weights_dtype = dtype_str(cfg.get("topk_weights_dtype"), "fp32")

    block_size_m = int(cfg.get("block_size_m", 64))
    block_size_n = int(cfg.get("block_size_n", 64))
    block_size_k = int(cfg.get("block_size_k", 32))
    group_size_m = int(cfg.get("group_size_m", 4))
    top_k = int(cfg.get("top_k", 2))
    naive_block_assignment = bool(cfg.get("naive_block_assignment", False))
    mul_routed_weight = bool(cfg.get("mul_routed_weight", True))
    compute_type = _TL_DTYPES[dtype_str(cfg.get("compute_type"), "bf16")]
    use_fp8_w8a8 = bool(cfg.get("use_fp8_w8a8", False))
    use_int8_w8a8 = bool(cfg.get("use_int8_w8a8", False))
    use_int8_w8a16 = bool(cfg.get("use_int8_w8a16", False))
    per_channel_quant = bool(cfg.get("per_channel_quant", False))
    has_bias = bool(cfg.get("has_bias", False))

    def p() -> Any:
        return _t("i64", 1)

    args = [
        _t(a_dtype, _M * _K),                       # a_ptr
        _t(b_dtype, _E * _N * _K),                  # b_ptr
        _t(b_bias_dtype, _E * _N),                  # b_bias_ptr
        _t(a_scale_dtype, 1),                       # a_scale_ptr
        _t(b_scale_dtype, 1),                       # b_scale_ptr
        _t(topk_weights_dtype, _M),                 # topk_weights_ptr
        _t("i32", _M),                              # sorted_token_ids_ptr
        _t("i32", _E * 4),                          # expert_ids_ptr
        _t("i32", 1),                               # num_tokens_post_padded_ptr
        p(), p(), p(), p(),                         # N_ptr, K_ptr, EM_ptr, num_valid_tokens_ptr
        p(), 1, p(), 1, p(), p(), 1,                # stride_am_ptr, stride_ak=1, stride_be_ptr, stride_bk=1, stride_bn_ptr, stride_cm_ptr, stride_cn=1
        p(), p(), p(), p(), p(), p(), p(),          # stride_asm/ask/bse/bsk/bsn/bbe/bbn ptrs
        0, 0, naive_block_assignment,               # group_n, group_k, naive_block_assignment (constexpr)
        block_size_m, block_size_n, block_size_k,   # BLOCK_SIZE_M/N/K (constexpr)
        group_size_m, 1, mul_routed_weight, top_k,  # GROUP_SIZE_M, SPLIT_K=1, MUL_ROUTED_WEIGHT, top_k (constexpr)
        compute_type,                               # compute_type (constexpr)
        use_fp8_w8a8, use_int8_w8a8, use_int8_w8a16, per_channel_quant, has_bias,  # (constexpr)
        _t(c_dtype, _M * _N),                       # c_ptr
    ]
    return args, {}
